File: poker-hands/typescript/src/Hand.ts
import { Card, card } from './Card.js';
import { Rank } from './Rank.js';
import { Suit } from './Suit.js';
import { HandRank } from './HandRank.js';
import { Compare } from './Compare.js';
import { InvalidHandError } from './InvalidHandError.js';

const HAND_SIZE = 5;

const RANK_CODES: Record<string, Rank> = {
  '2': Rank.Two, '3': Rank.Three, '4': Rank.Four, '5': Rank.Five,
  '6': Rank.Six, '7': Rank.Seven, '8': Rank.Eight, '9': Rank.Nine,
  T: Rank.Ten, J: Rank.Jack, Q: Rank.Queen, K: Rank.King, A: Rank.Ace,
};

const SUIT_CODES: Record<string, Suit> = {
  C: Suit.Clubs, D: Suit.Diamonds, H: Suit.Hearts, S: Suit.Spades,
};

export class Hand {
  readonly cards: readonly Card[];

  constructor(cards: readonly Card[]) {
    if (cards.length !== HAND_SIZE) {
      throw new InvalidHandError(`A hand must have exactly 5 cards (got ${cards.length})`);
    }
    this.cards = [...cards];
  }

  /**
   * Parse a hand from shorthand notation like "2H 3D 5S 9C KD".
   * Each token is a two-character card: rank code (2-9, T, J, Q, K, A) + suit code (C, D, H, S).
   */
  static parse(shorthand: string): Hand {
    const tokens = shorthand.trim().split(/\s+/).filter((t) => t.length > 0);
    const cards = tokens.map((token) => parseCard(token));
    return new Hand(cards);
  }

  evaluate(): HandRank {
    const groups = rankGroupsByCountDescending(this.cards);
    const counts = groups.map((g) => g.count);
    const isFlush = new Set(this.cards.map((c) => c.suit)).size === 1;
    const isStraight = computeIsStraight(this.cards);

    if (isStraight && isFlush) return HandRank.StraightFlush;
    if (counts[0] === 4) return HandRank.FourOfAKind;
    if (counts[0] === 3 && counts[1] === 2) return HandRank.FullHouse;
    if (isFlush) return HandRank.Flush;
    if (isStraight) return HandRank.Straight;
    if (counts[0] === 3) return HandRank.ThreeOfAKind;
    if (counts[0] === 2 && counts[1] === 2) return HandRank.TwoPair;
    if (counts[0] === 2) return HandRank.Pair;
    return HandRank.HighCard;
  }

  compareTo(other: Hand): Compare {
    const myRank = this.evaluate();
    const theirRank = other.evaluate();
    if (myRank > theirRank) return Compare.Player1Wins;
    if (myRank < theirRank) return Compare.Player2Wins;

    const mySig = tieBreakSignature(this.cards);
    const theirSig = tieBreakSignature(other.cards);
    for (let i = 0; i < mySig.length; i++) {
      const a = mySig[i]!;
      const b = theirSig[i]!;
      if (a > b) return Compare.Player1Wins;
      if (a < b) return Compare.Player2Wins;
    }
    return Compare.Tie;
  }
}

function parseCard(token: string): Card {
  if (token.length !== 2) {
    throw new InvalidHandError(`Invalid card token '${token}'`);
  }
  const rank = RANK_CODES[token[0]!];
  if (rank === undefined) throw new InvalidHandError(`Invalid rank code '${token[0]}'`);
  const suit = SUIT_CODES[token[1]!];
  if (suit === undefined) throw new InvalidHandError(`Invalid suit code '${token[1]}'`);
  return card(rank, suit);
}

interface RankGroup { rank: Rank; count: number; }

function rankGroupsByCountDescending(cards: readonly Card[]): RankGroup[] {
  const counts = new Map<Rank, number>();
  for (const c of cards) counts.set(c.rank, (counts.get(c.rank) ?? 0) + 1);
  return [...counts.entries()]
    .map(([rank, count]) => ({ rank, count }))
    .sort((a, b) => (b.count - a.count) || (b.rank - a.rank));
}

/**
 * Canonical tie-break signature: ranks ordered first by group size (descending),
 * then by rank (descending). A positional compare correctly implements every
 * same-rank tie-breaker rule in SCENARIOS.md.
 */
function tieBreakSignature(cards: readonly Card[]): Rank[] {
  const signature: Rank[] = [];
  for (const g of rankGroupsByCountDescending(cards)) {
    for (let i = 0; i < g.count; i++) signature.push(g.rank);
  }
  return signature;
}

function computeIsStraight(cards: readonly Card[]): boolean {
  const sorted = cards.map((c) => c.rank as number).sort((a, b) => a - b);
  for (let i = 1; i < sorted.length; i++) {
    if (sorted[i]! !== sorted[i - 1]! + 1) return false;
  }
  return true;
}

File: poker-hands/python/src/poker_hands/hand.py
from collections import Counter
from typing import Iterable, List, Tuple

from .card import Card
from .compare import Compare
from .exceptions import InvalidHandError
from .hand_rank import HandRank
from .rank import Rank
from .suit import Suit

HAND_SIZE = 5

_RANK_CODES = {
    "2": Rank.TWO, "3": Rank.THREE, "4": Rank.FOUR, "5": Rank.FIVE,
    "6": Rank.SIX, "7": Rank.SEVEN, "8": Rank.EIGHT, "9": Rank.NINE,
    "T": Rank.TEN, "J": Rank.JACK, "Q": Rank.QUEEN, "K": Rank.KING, "A": Rank.ACE,
}

_SUIT_CODES = {
    "C": Suit.CLUBS, "D": Suit.DIAMONDS, "H": Suit.HEARTS, "S": Suit.SPADES,
}


class Hand:
    def __init__(self, cards: Iterable[Card]) -> None:
        cards_list = list(cards)
        if len(cards_list) != HAND_SIZE:
            raise InvalidHandError(
                f"A hand must have exactly 5 cards (got {len(cards_list)})"
            )
        self._cards: Tuple[Card, ...] = tuple(cards_list)

    @property
    def cards(self) -> Tuple[Card, ...]:
        return self._cards

    @classmethod
    def from_string(cls, shorthand: str) -> "Hand":
        """Parse a hand from shorthand notation like "2H 3D 5S 9C KD".

        Each token is a two-character card: rank code (2-9, T, J, Q, K, A)
        plus suit code (C, D, H, S).
        """
        tokens = shorthand.split()
        return cls(_parse_card(token) for token in tokens)

    def evaluate(self) -> HandRank:
        groups = _rank_groups_by_count_descending(self._cards)
        counts = [count for _, count in groups]
        is_flush = len({c.suit for c in self._cards}) == 1
        is_straight = _is_straight(self._cards)

        if is_straight and is_flush:
            return HandRank.STRAIGHT_FLUSH
        if counts[0] == 4:
            return HandRank.FOUR_OF_A_KIND
        if counts[0] == 3 and len(counts) > 1 and counts[1] == 2:
            return HandRank.FULL_HOUSE
        if is_flush:
            return HandRank.FLUSH
        if is_straight:
            return HandRank.STRAIGHT
        if counts[0] == 3:
            return HandRank.THREE_OF_A_KIND
        if counts[0] == 2 and len(counts) > 1 and counts[1] == 2:
            return HandRank.TWO_PAIR
        if counts[0] == 2:
            return HandRank.PAIR
        return HandRank.HIGH_CARD

    def compare_to(self, other: "Hand") -> Compare:
        my_rank = self.evaluate()
        their_rank = other.evaluate()
        if my_rank > their_rank:
            return Compare.PLAYER1_WINS
        if my_rank < their_rank:
            return Compare.PLAYER2_WINS

        my_sig = _tie_break_signature(self._cards)
        their_sig = _tie_break_signature(other.cards)
        for mine, theirs in zip(my_sig, their_sig):
            if mine > theirs:
                return Compare.PLAYER1_WINS
            if mine < theirs:
                return Compare.PLAYER2_WINS
        return Compare.TIE


def _parse_card(token: str) -> Card:
    if len(token) != 2:
        raise InvalidHandError(f"Invalid card token '{token}'")
    rank = _RANK_CODES.get(token[0])
    if rank is None:
        raise InvalidHandError(f"Invalid rank code '{token[0]}'")
    suit = _SUIT_CODES.get(token[1])
    if suit is None:
        raise InvalidHandError(f"Invalid suit code '{token[1]}'")
    return Card(rank, suit)


def _rank_groups_by_count_descending(cards: Tuple[Card, ...]) -> List[Tuple[Rank, int]]:
    counts = Counter(c.rank for c in cards)
    return sorted(counts.items(), key=lambda g: (-g[1], -g[0]))


def _tie_break_signature(cards: Tuple[Card, ...]) -> List[Rank]:
    """Canonical tie-break signature: ranks ordered first by group size (descending),
    then by rank (descending). A positional compare correctly implements every
    same-rank tie-breaker rule in SCENARIOS.md.
    """
    signature: List[Rank] = []
    for rank, count in _rank_groups_by_count_descending(cards):
        signature.extend([rank] * count)
    return signature


def _is_straight(cards: Tuple[Card, ...]) -> bool:
    ordered = sorted(int(c.rank) for c in cards)
    return all(ordered[i] == ordered[i - 1] + 1 for i in range(1, len(ordered)))
